lib/asyncSender.ts

```diff
@@ -0,0 +1,181 @@
+        const promise = this.deliveryDispositionMap.get(id) as PromiseLike;
+        clearTimeout(promise.timer);
+        log.sender(
+          "[%s] Event: 'Accepted', Found the delivery with id %d in the map of " +
+          "sender '%s' on amqp session '%s' and cleared the timer.",
+          this.connection.id, id, this.name, this.session.id
+        );
+        const deleteResult = this.deliveryDispositionMap.delete(id);
+        log.sender(
+          "[%s] Event: 'Accepted', Successfully deleted the delivery with id %d from " +
+          "the map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+          this.connection.id, id, this.name, this.session.id, deleteResult
+        );
+        return promise.resolve(context.delivery);
+      }
+    };
+
+    const onFailure = (eventName: string, id: number, error?: Error) => {
+      if (this.deliveryDispositionMap.has(id)) {
+        log.sender(
+          "[%s] Event: '%s', Found the delivery with id %d in the map of " +
+          "sender '%s' on amqp session '%s' and cleared the timer.",
+          this.connection.id, eventName, id, this.name, this.session.id
+        );
+        const promise = this.deliveryDispositionMap.get(id) as PromiseLike;
+        clearTimeout(promise.timer);
+        const deleteResult = this.deliveryDispositionMap.delete(id);
+        log.sender(
+          "[%s] Event: '%s', Successfully deleted the delivery with id %d from the " +
+          " map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+          this.connection.id, eventName, id, this.name, this.session.id, deleteResult
+        );
+        const msg = `Sender '${this.name}' on amqp session '${this.session.id}', received a ` +
+          `'${eventName}' disposition. Hence we are rejecting the promise.`;
+        const err = error || new Error(msg);
+        log.sender("[%s] %s", this.connection.id, msg);
+        return promise.reject(err);
+      }
+    };
+
+    const onError = (eventName: string, error?: Error) => {
+      for (const id of this.deliveryDispositionMap.keys()) {
+        onFailure(eventName, id, error);
+      }
+    };
+
+    this.on(SenderEvents.accepted, onSuccess);
+    this.on(SenderEvents.rejected, (context: EventContext) => {
+      onFailure(SenderEvents.rejected, context.delivery!.id, context!.delivery!.remote_state!.error);
+    });
+    this.on(SenderEvents.released, (context: EventContext) => {
+      onFailure(SenderEvents.released, context.delivery!.id, context!.delivery!.remote_state!.error);
+    });
+    this.on(SenderEvents.modified, (context: EventContext) => {
+      onFailure(SenderEvents.modified, context.delivery!.id, context!.delivery!.remote_state!.error);
+    });
+
+    // The user may have it's custom reconnect logic for bringing the sender link back online and
+    // retry logic for sending messages on failures hence they can provide their error handlers
```

Comment:
Assuming the user has the custom reconnect logic and brings back the sender, would the server know that this is the same link to send the `accepted` or `rejected` events?

Reply:
I was thinking if in SB and EH, we need to use this AsyncSender, then how will we do it. Today we have our handlers for `*_error` and `*close` on the sender and we have our own reconnect logic that takes care of bringing the connection online and recreating things on top of it. If we clear all the promises here, then we may unnecessarily screw up things for libraries like SB. Hence kept it simple.

> Assuming the user has the custom reconnect logic and brings back the sender, would the server know that this is the same link to send the accepted or rejected events?

Your question is correct. I am thinking (writing) out loud about how do we handle this today in SB? It is possible that connection gets disconnected while the sender is sending messages. Promise for the current message will timeout and the retry for sending message will take care of making the next attempt to send the message.
Now, one of the two scenarios can happen when the sender is waiting on the Promise for the current send operation and the connection goes down.
1. The promise times out and if the connection was brought back up while the sender was waiting for the promise to resolve/reject, it will retry one more time to send the message
2. The promise times out and the send figures that the sender is not open, which should trigger the activity to bring either the sender or things under it back up and then the send operation will be attempted one more time.

It is possible that many users of rhea-promise will not have the reconnect logic that we have in SB. Hence have kept a simple option to clear out disposition map and reject all the promises with an appropriate error message.